openoa/utils/downloader.py

```diff
@@ -482,3 +727,182 @@ def get_merra2_monthly(
+        end_date(:obj:`str`): The final year, month, and day that data is downloaded for. This should be
+            provided as a string in the format "YYYY-MM-DD". Defaults to current date. Note that data
+            may not be available yet for the most recent couple months.
+        calc_derived_vars (:obj:`bool`, optional): Boolean that specifies whether wind speed, wind
+            direction, and air density are computed from the downloaded reanalysis variables and
+            saved. Defaults to False.
+
+    Returns:
+        df(:obj:`dataframe`): A dataframe containing time series of the requested reanalysis
+            variables:
+            1. u_ms: the U component of wind speed at a height of 50 m in m/s.
+            2. v_ms: the V component of wind speed at a height of 50 m in m/s.
+            3. temperature_K: air temperature at a height of 2 m in Kelvin.
+            4. surf_pres_Pa: surface pressure in Pascals.
+
+    Raises:
+        ValueError: If the start_year is greater than the end_year.
+    """
+
+    logger.info("Please note access to MERRA2 data requires registration")
+    logger.info("Please see: https://disc.gsfc.nasa.gov/data-access#python-requests")
+
+    # base url containing the monthly data set M2T1NXSLV
+    base_url = r"https://goldsmr4.gesdisc.eosdis.nasa.gov/opendap/MERRA2/M2T1NXSLV.5.12.4/"
+
+    # create save_pathname if it does not exist
+    save_pathname = Path(save_pathname).resolve()
+    if not save_pathname.exists():
+        save_pathname.mkdir()
+
+    # get the current date
+    now = datetime.datetime.now()
+
+    # assign end_year to current year if not provided by the user
+    if end_date is None:
+        end_date = f"{now.year}-{now.month:02}-{now.day:02}"
+
+    # convert dates to datetime objects
+    start_date = datetime.datetime.strptime(start_date, "%Y-%m-%d")
+    end_date = datetime.datetime.strptime(end_date, "%Y-%m-%d")
+    end_date += datetime.timedelta(hours=23, minutes=59)  # include all times in last day
+
+    # check that the start and end dates are the right way around
+    if start_date > end_date:
+        logger.error("The start_date should be less than or equal to the end_date")
+        logger.error(f"start_date = {start_date.date()}, end_date = {end_date.date()}")
+        raise ValueError("The start_date should be less than or equal to the end_date")
+
+    # list all years that will be downloaded
+    years = list(range(start_date.year, end_date.year + 1, 1))
+
+    # download the data
+    for year in years:
+        # limit to months of interest
+        if year == start_date.year:
+            if year == end_date.year:
+                months = list(range(start_date.month, end_date.month + 1, 1))
+            else:
+                months = list(range(start_date.month, 12 + 1, 1))
+        elif year == end_date.year:
+            months = list(range(1, end_date.month + 1, 1))
+        else:
+            months = list(range(1, 12 + 1, 1))
```

Comment:
The extra 1 at the end is the default, even when the starting point is changed. Though if that's just a style thing, then it doesn't bother me, but thought I'd flag it just in case.